### PR TITLE
Fix recursion performance

### PR DESCRIPTION
**Pull request by gaurav-arya**
This PR simplifies the recursive structure of walking a functor. It uses a trick from https://github.com/JuliaLang/julia/issues/47760 to resolve https://github.com/FluxML/Functors.jl/issues/59. 

The functionality of `fmap` after this PR is identical to its previous one. Thus, this PR lays bare the fact that the argument `f` is ultimately useless for the version of `fmap` that takes a user-specified walk. Note that this was public API, however, so I kept it in the signature and added an explanatory note.

Obviously the `fmap` API is highly imperfect, but I think this PR is a strict improvement, that simply lays bare some of the issues. (Ultimately, I think that the use of `fmap` with a user-specified arbitrary walk should probably be deprecated: walks are just way too flexible to allow arbitrary walks to be paired with `fmap`. The semantics of `fmap` seem to match best with the specific use case of `ExcludeWalk`: `ExcludeWalk` is what actually uses `f`, after all. The general function allowing for arbitrary walks should probably be called `recursivewalk` or something. But that would be very breaking.)

### CODE REVIEW

**Review by darsnack — APPROVED** [review]
Are there packages outside Flux using the walk-based API? Better to make a breaking change soon before more people adopt the new (bad) API.

File: src/maps.jl [hunk 1]
```diff
@@ -1,4 +1,20 @@
...
+this API (issue #62).
+=#
+function fmap(walk::AbstractWalk, f, x, ys...)
+  #=
+  The below construct avoids a performance penalty
+  for recursive constructs in an anonymous function.
+  See Julia issue #47760 and Functors.jl issue #59.
+  =#
+  recurse(xs...) = walk(var"#self#", xs...)
+  walk(recurse, x, ys...)
```
Comment (ToucheSir) [inline src/maps.jl:16]:
```suggestion
  # This avoids a performance penalty for recursive constructs in an anonymous function.
  # See Julia issue #47760 and Functors.jl issue #59.
  recurse(xs...) = walk(var"#self#", xs...)
  walk(recurse, x, ys...)
```

File: src/maps.jl [hunk 2]
```diff
@@ -1,4 +1,20 @@
-fmap(walk::AbstractWalk, f, x, ys...) = walk((xs...) -> fmap(walk, f, xs...), x, ys...)
+#=
+Note that the argument f is not actually used in 
+the below version of fmap, which takes a user-specified
+walk. Since this function is public API, the f argument
+has been left in the signature until a (likely breaking)
+decision can be made in the future about how to clean up
+this API (issue #62).
+=#
```
Comment (ToucheSir) [inline src/maps.jl:8]:
```suggestion
# Note that the argument f is not actually used in this method.
# See issue #62 for a discussion on how best to remove it.
```

**Review by ToucheSir — APPROVED** [review]
Thanks! Just a couple of formatting nits from me.

**Comment by gaurav-arya** [comment]
I don't believe this has got a release?

**Comment by ToucheSir** [comment]
I don't remember why. Might've held off because of the recent TagBot issues that prevented tags and release notes from being published. See https://github.com/JuliaRegistries/General/pull/77862

**Pull request merged**